tls/s2n_x509_validator.c

```diff
@@ -391,6 +427,11 @@ S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *val
         validator->state = VALIDATED;
     }
 
+    DEFER_CLEANUP(struct s2n_pkey public_key = {0}, s2n_pkey_free);
+    s2n_pkey_zero_init(&public_key);
+    s2n_parsed_extensions_list first_certificate_extensions = {0};
+    RESULT_GUARD(s2n_read_leaf_info(conn, cert_chain_in, cert_chain_len, &public_key, pkey_type, &first_certificate_extensions));
+
```

Comment:
Is parsing the first cert twice necessary? Or is it just super cheap and this is cleaner?

Reply:
Yeah, it isn't necessary, though I think it is cleaner. The problem is the INIT stage is intended to be skipped the second time `s2n_x509_validator_verify_cert_chain` is called, after the async callback becomes unblocked. So, a solution could be to do an allocation in the s2n_x509_validator in this stage to store `public_key` and `first_certificate_extensions`, which could then be retrieved in the second call. However, since reading the first certificate off the stuffer twice is cheap, I figured this way is nicer.

Reply:
Oh right: the extensions could be processed the first time easily enough, but `public_key` is much trickier.

Is it a problem that you're now processing the extensions after the certificate chain is validated? The current Certificate message extensions support OCSP and SCT, so I'd expect them to be involved in validation / occur when CRL validation occurs.

Reply:
Understood. And, just to confirm, you're processing the extensions at the same point in the validation process as where they were processed before?

Reply:
Correct. Previously, the extensions were parsed and the leaf certificate extensions were saved at the beginning: https://github.com/aws/s2n-tls/blob/f29843579ada9eeccffa869f4b088fddd4231716/tls/s2n_x509_validator.c#L343-L352

But the processing of the leaf certificate extensions occurred after all of the X509_verify_cert call: https://github.com/aws/s2n-tls/blob/f29843579ada9eeccffa869f4b088fddd4231716/tls/s2n_x509_validator.c#L394-L396

With the refactor, the processing still occurs at the end in the same place.